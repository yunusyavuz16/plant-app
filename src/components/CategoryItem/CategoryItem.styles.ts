import { colors, layout, spacing, typography } from "@/constants/theme";
import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  categoryCard: {
    flex: 1,
    paddingStart: spacing.lg,
    paddingTop: spacing.lg,
    borderRadius: layout.radius.lg,
    backgroundColor: colors.background.lightGreen,
    borderWidth: 1,
    borderColor: colors.border.green,
    justifyContent: "space-between",
    minHeight: layout.card.categoryHeight,
  },
  categoryTitle: {
    fontSize: typography.size.lg,
    fontWeight: "600",
    textAlign: "left",
    marginBottom: spacing.sm,
  },
  categoryImage: {
    width: "100%",
    height: 100,
    borderRadius: layout.radius.md,
  },
});
