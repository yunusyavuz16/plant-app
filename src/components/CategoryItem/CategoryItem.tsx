import { Image } from "expo-image";
import React, { memo } from "react";
import { Text, TouchableOpacity } from "react-native";
import { styles } from "./CategoryItem.styles";

const CategoryItem = memo(
  ({ title, url }: { title: string; url: string }) => {
    return (
      <TouchableOpacity style={styles.categoryCard}>
        <Text style={styles.categoryTitle}>{title}</Text>
        <Image
          cachePolicy="memory-disk"
          source={{ uri: url }}
          style={styles.categoryImage}
        />
      </TouchableOpacity>
    );
  }
);

export default CategoryItem;
