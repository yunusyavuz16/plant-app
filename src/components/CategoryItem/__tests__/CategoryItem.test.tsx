import React from 'react';
import renderer, { act } from 'react-test-renderer';
import CategoryItem from '../CategoryItem';

// Mock the styles before importing the component
jest.mock('../CategoryItem.styles', () => ({
  styles: {
    categoryCard: {
      flex: 1,
      paddingStart: 16,
      paddingTop: 16,
      borderRadius: 12,
      backgroundColor: '#F5F9F6',
      borderWidth: 1,
      borderColor: '#E5EBE6',
      justifyContent: 'space-between',
      minHeight: 180,
    },
    categoryTitle: {
      fontSize: 18,
      fontWeight: '600',
      textAlign: 'left',
      marginBottom: 8,
    },
    categoryImage: {
      width: '100%',
      height: 100,
      borderRadius: 8,
    },
  },
}));

// Mock React Native components
jest.mock('react-native', () => ({
  TouchableOpacity: 'TouchableOpacity',
  Text: 'Text',
  StyleSheet: {
    create: (styles: any) => styles,
  },
}));

// Mock the expo-image component
jest.mock('expo-image', () => ({
  Image: 'Image',
  ImageContentFit: {
    cover: 'cover',
    contain: 'contain',
  },
}));

describe('CategoryItem', () => {
  const defaultProps = {
    title: 'Test Category',
    url: 'https://example.com/test.jpg',
  };

  const renderComponent = (props = {}) => {
    let component: renderer.ReactTestRenderer;
    act(() => {
      component = renderer.create(<CategoryItem {...defaultProps} {...props} />);
    });
    return component!;
  };

  it('renders correctly with required props', () => {
    const tree = renderComponent().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles onPress callback', () => {
    const onPressMock = jest.fn();
    const component = renderComponent({ onPress: onPressMock });

    const touchable = component.root.findByProps({ testID: 'category-item' });
    act(() => {
      touchable.props.onPress();
    });

    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it('does not call onPress when no handler is provided', () => {
    const component = renderComponent();
    const touchable = component.root.findByProps({ testID: 'category-item' });

    // Should not throw when pressed
    act(() => {
      touchable.props.onPress();
    });
  });

  it('handles long titles gracefully', () => {
    const longTitle = 'This is a very long category title that might cause layout issues if not handled properly';
    const tree = renderComponent({ title: longTitle }).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with correct styles', () => {
    const component = renderComponent();
    const touchable = component.root.findByProps({ testID: 'category-item' });
    const title = component.root.findByProps({ children: defaultProps.title });

    expect(touchable.props.style[0]).toEqual(expect.objectContaining({
      flex: 1,
      paddingStart: 16,
      paddingTop: 16,
      borderRadius: 12,
      backgroundColor: '#F5F9F6',
      borderWidth: 1,
      borderColor: '#E5EBE6',
      justifyContent: 'space-between',
      minHeight: 180,
    }));

    expect(title.props.style[0]).toEqual(expect.objectContaining({
      fontSize: 18,
      fontWeight: '600',
      textAlign: 'left',
      marginBottom: 8,
    }));
  });

  it('applies custom styles correctly', () => {
    const customStyle = {
      backgroundColor: '#FF0000',
      padding: 20,
    };
    const customTitleStyle = {
      fontSize: 24,
      color: '#0000FF',
    };

    const component = renderComponent({
      style: customStyle,
      titleStyle: customTitleStyle,
    });

    const touchable = component.root.findByProps({ testID: 'category-item' });
    const title = component.root.findByProps({ children: defaultProps.title });

    expect(touchable.props.style).toEqual([
      expect.objectContaining({
        flex: 1,
        paddingStart: 16,
        paddingTop: 16,
      }),
      customStyle,
    ]);

    expect(title.props.style).toEqual([
      expect.objectContaining({
        fontSize: 18,
        fontWeight: '600',
      }),
      customTitleStyle,
    ]);
  });

  it('handles disabled state correctly', () => {
    const onPressMock = jest.fn();
    const component = renderComponent({
      disabled: true,
      onPress: onPressMock,
    });

    const touchable = component.root.findByProps({ testID: 'category-item' });
    expect(touchable.props.disabled).toBe(true);
    expect(touchable.props.accessibilityState).toEqual({ disabled: true });

    act(() => {
      touchable.props.onPress();
    });
    expect(onPressMock).not.toHaveBeenCalled();
  });

  it('sets correct accessibility properties', () => {
    const customAccessibilityLabel = 'Custom Category Label';
    const component = renderComponent({
      accessibilityLabel: customAccessibilityLabel,
    });

    const touchable = component.root.findByProps({ testID: 'category-item' });
    const title = component.root.findByProps({ children: defaultProps.title });
    const image = component.root.findByProps({ accessibilityRole: 'image' });

    expect(touchable.props.accessibilityLabel).toBe(customAccessibilityLabel);
    expect(touchable.props.accessibilityRole).toBe('button');
    expect(title.props.accessibilityRole).toBe('header');
    expect(image.props.accessibilityLabel).toBe(`${defaultProps.title} category image`);
  });

  it('applies correct image properties', () => {
    const component = renderComponent({
      imageFit: 'contain',
    });

    const image = component.root.findByProps({ accessibilityRole: 'image' });
    expect(image.props.contentFit).toBe('contain');
    expect(image.props.transition).toBe(200);
    expect(image.props.cachePolicy).toBe('memory-disk');
  });
});