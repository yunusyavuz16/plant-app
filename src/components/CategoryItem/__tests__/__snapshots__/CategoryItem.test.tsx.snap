// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem handles long titles gracefully 1`] = `
<TouchableOpacity
  accessibilityLabel="This is a very long category title that might cause layout issues if not handled properly"
  accessibilityRole="button"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  disabled={false}
  onPress={[Function]}
  style={
    [
      {
        "backgroundColor": "#F5F9F6",
        "borderColor": "#E5EBE6",
        "borderRadius": 12,
        "borderWidth": 1,
        "flex": 1,
        "justifyContent": "space-between",
        "minHeight": 180,
        "paddingStart": 16,
        "paddingTop": 16,
      },
      undefined,
    ]
  }
  testID="category-item"
>
  <Text
    accessibilityRole="header"
    ellipsizeMode="tail"
    numberOfLines={2}
    style={
      [
        {
          "fontSize": 18,
          "fontWeight": "600",
          "marginBottom": 8,
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    This is a very long category title that might cause layout issues if not handled properly
  </Text>
  <Image
    accessibilityLabel="This is a very long category title that might cause layout issues if not handled properly category image"
    accessibilityRole="image"
    cachePolicy="memory-disk"
    contentFit="cover"
    source={
      {
        "uri": "https://example.com/test.jpg",
      }
    }
    style={
      {
        "borderRadius": 8,
        "height": 100,
        "width": "100%",
      }
    }
    transition={200}
  />
</TouchableOpacity>
`;

exports[`CategoryItem renders correctly with required props 1`] = `
<TouchableOpacity
  accessibilityLabel="Test Category"
  accessibilityRole="button"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  disabled={false}
  onPress={[Function]}
  style={
    [
      {
        "backgroundColor": "#F5F9F6",
        "borderColor": "#E5EBE6",
        "borderRadius": 12,
        "borderWidth": 1,
        "flex": 1,
        "justifyContent": "space-between",
        "minHeight": 180,
        "paddingStart": 16,
        "paddingTop": 16,
      },
      undefined,
    ]
  }
  testID="category-item"
>
  <Text
    accessibilityRole="header"
    ellipsizeMode="tail"
    numberOfLines={2}
    style={
      [
        {
          "fontSize": 18,
          "fontWeight": "600",
          "marginBottom": 8,
          "textAlign": "left",
        },
        undefined,
      ]
    }
  >
    Test Category
  </Text>
  <Image
    accessibilityLabel="Test Category category image"
    accessibilityRole="image"
    cachePolicy="memory-disk"
    contentFit="cover"
    source={
      {
        "uri": "https://example.com/test.jpg",
      }
    }
    style={
      {
        "borderRadius": 8,
        "height": 100,
        "width": "100%",
      }
    }
    transition={200}
  />
</TouchableOpacity>
`;
